warehouse/router/tracker.go

```diff
@@ -10,196 +10,170 @@ import (
 			}
 		}
 
 		nextExecTime := execTime.Add(r.config.uploadStatusTrackFrequency)
 		select {
 		case <-ctx.Done():
-			r.logger.Infon("context is cancelled, stopped running tracking")
+			r.logger.Infon("Context cancelled. Exiting cron tracker")
 			return nil
 		case <-time.After(time.Until(nextExecTime)):
 		}
 	}
 }
 
-// Track tracks the status of the warehouse uploads for the corresponding cases:
-// 1. Staging files is not picked.
-// 2. Upload job is struck
-func (r *Router) Track(
-	ctx context.Context,
-	warehouse *model.Warehouse,
-	config *config.Config,
-) error {
-	var (
-		createdAt         sql.NullTime
-		exists            bool
-		syncFrequency     = "1440"
-		now               = timeutil.Now
-		nowSQL            = "NOW()"
-		failedStatusRegex = "%_failed"
-		timeWindow        = config.GetDuration("Warehouse.uploadBufferTimeInMin", 180, time.Minute)
-		source            = warehouse.Source
-		destination       = warehouse.Destination
-	)
-
-	if r.nowSQL != "" {
-		nowSQL = r.nowSQL
-	}
-	if r.now != nil {
-		now = r.now
+func (r *Router) retryTrackSync(ctx context.Context, warehouse *model.Warehouse) error {
+	o := func() error {
+		return r.trackSync(ctx, warehouse)
 	}
+	b := backoff.WithContext(
+		backoff.WithMaxRetries(
+			backoff.NewExponentialBackOff(),
+			uint64(r.config.cronTrackerRetries.Load()),
+		),
+		ctx,
+	)
+	return backoff.Retry(o, b)
+}
 
-	trackUploadMissingStat := r.statsFactory.NewTaggedStat("warehouse_track_upload_missing", stats.GaugeType, stats.Tags{
-		"workspaceId":   warehouse.WorkspaceID,
-		"module":        moduleName,
-		"destType":      r.destType,
-		"sourceId":      source.ID,
-		"destinationId": destination.ID,
-		"warehouseID": misc.GetTagName(
-			destination.ID,
-			source.Name,
-			destination.Name,
-			misc.TailTruncateStr(source.ID, 6)),
-	})
-	trackUploadMissingStat.Gauge(0)
-
-	if !source.Enabled || !destination.Enabled {
+func (r *Router) trackSync(ctx context.Context, warehouse *model.Warehouse) error {
+	if !warehouse.IsEnabled() || r.isWithinExcludeWindow(warehouse) {
 		return nil
 	}
 
-	excludeWindow := warehouse.GetMapDestinationConfig(model.ExcludeWindowSetting)
-	excludeWindowStartTime, excludeWindowEndTime := excludeWindowStartEndTimes(excludeWindow)
-	if checkCurrentTimeExistsInExcludeWindow(now(), excludeWindowStartTime, excludeWindowEndTime) {
+	createdAt, err := r.getOldestStagingFile(ctx, warehouse)
+	if err != nil {
+		return err
+	}
+	if createdAt.IsZero() {
 		return nil
 	}
 
-	if sf := warehouse.GetStringDestinationConfig(r.conf, model.SyncFrequencySetting); sf != "" {
-		syncFrequency = sf
-	}
-	if value, err := strconv.Atoi(syncFrequency); err == nil {
-		timeWindow += time.Duration(value) * time.Minute
+	exists, err := r.checkUploadStatus(ctx, warehouse, createdAt)
+	if err != nil {
+		return err
 	}
 
+	r.recordUploadMissingMetric(warehouse, exists)
+	return nil
+}
+
+func (r *Router) isWithinExcludeWindow(warehouse *model.Warehouse) bool {
```

Comment:
Would it be more appropriate to define these new methods, such as `getOldestStagingFile` or `isWithinExcludeWindow`, on a separate cron struct instead of adding them to the `Router`? These methods seem to be unrelated to the core responsibilities of the Router.